src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/LegacyCommonAnnotatedSecurityAccessKey.cs

```diff
@@ -29,15 +30,22 @@ public override Tuple<string, string> GetMatchIdAndName(string match)
                 return null;
             }
 #endif
-            string providerSignature = match.Substring(LegacyCommonAnnotatedSecurityKey.ProviderFixedSignatureOffset, 4);
-            if (providerSignature != ProviderSignature)
+            if (string.CompareOrdinal(match, LegacyCommonAnnotatedSecurityKey.ProviderFixedSignatureOffset, ProviderSignature, 0, 4) != 0)
             {
                 return null;
             }
 
-            return match[LegacyCommonAnnotatedSecurityKey.StandardFixedSignatureOffset + 5] == '9'
-                ? new Tuple<string, string>(Id, Name)
-                : null;
+            if (string.CompareOrdinal(match, LegacyCommonAnnotatedSecurityKey.ProviderReservedOffset, ProviderData, 0, 4) != 0)
+            {
+                return null;
+            }
+
+            if (match[LegacyCommonAnnotatedSecurityKey.StandardFixedSignatureOffset + 5] != '9')
```

Comment:
A comment here would be good. It's subtle that you see the 9 in pattern above but need to check it here in high-perf case.